src/lib/string_pool.rs

```diff
@@ -0,0 +1,506 @@
+
+#[cfg(feature = "mozilla")]
+const fn init_block_size_const() -> usize {
+    // FIXME: should be `offset_of(BLOCK, s)`, but that's not supported yet,
+    // so we over-estimate its offset
+    1024 - (std::mem::size_of::<BLOCK>() / std::mem::size_of::<XML_Char>())
+}
+
+#[cfg(not(feature = "mozilla"))]
+const fn init_block_size_const() -> usize {
+    1024
+}
+
+fn poolBytesToAllocateFor(mut blockSize: c_int) -> size_t {
+    /* Unprotected math would be:
+     ** return offsetof(BLOCK, s) + blockSize * sizeof(XML_Char);
+     **
+     ** Detect overflow, avoiding _signed_ overflow undefined behavior
+     ** For a + b * c we check b * c in isolation first, so that addition of a
+     ** on top has no chance of making us accept a small non-negative number
+     */
+    let stretch: size_t = ::std::mem::size_of::<XML_Char>() as c_ulong; /* can be 4 bytes */
+    if blockSize <= 0 {
+        return 0u64;
+    }
+    if blockSize > (INT_MAX as c_ulong).wrapping_div(stretch) as c_int {
+        return 0u64;
+    }
+    let stretchedBlockSize: c_int = blockSize * stretch as c_int;
+    let bytesToAllocate: c_int =
+        (12u64).wrapping_add(stretchedBlockSize as c_uint as c_ulong) as c_int;
+    if bytesToAllocate < 0 {
+        return 0u64;
+    }
+    bytesToAllocate as size_t
+}
+
+rental! {
+    mod rental_pool {
+        use super::*;
+
+        #[rental(debug)]
+        pub(crate) struct InnerStringPool {
+            bump: Box<Bump>,
+            current_bump_vec: RefCell<BumpVec<'bump, XML_Char>>,
+        }
+    }
+}
+
+use rental_pool::InnerStringPool;
+
+pub(crate) struct StringPool(Option<InnerStringPool>);
+
+impl StringPool {
+    pub(crate) fn try_new() -> Result<Self, ()> {
+        let bump = Bump::try_new().map_err(|_| ())?;
+        let boxed_bump = Box::try_new(bump).map_err(|_| ())?;
+
+        Ok(StringPool(Some(InnerStringPool::new(
+            boxed_bump,
+            |bump| RefCell::new(BumpVec::new_in(&bump)),
+        ))))
+    }
+
+    fn inner(&self) -> &InnerStringPool {
+        self.0.as_ref().unwrap()
+    }
+
+    /// Determines whether or not the current BumpVec is empty.
+    pub(crate) fn is_empty(&self) -> bool {
+        self.inner().rent(|vec| vec.borrow().is_empty())
+    }
+
+    /// Determines whether or not the current BumpVec is full.
+    pub(crate) fn is_full(&self) -> bool {
+        self.inner().rent(|vec| vec.borrow().len() == vec.borrow().capacity())
+    }
+
+    /// Gets the current vec, converts it into a slice, and resets
+    /// bookkeeping so that it will create a new vec next time.
+    pub(crate) fn consume_current_vec(&self) -> &mut [XML_Char] {
+        let pool = unsafe { self.inner().all_erased() };
+        let mut vec = BumpVec::new_in(&pool.bump);
+        pool.current_bump_vec.replace(vec).into_bump_slice_mut()
+    }
+
+    /// Resets the current bump vec to the beginning
+    pub(crate) fn clear_current(&self) {
+        self.set_len(0)
```

Comment:
`clear` will deallocate the contents, this is just resetting the buffer so that it can be reused. Maybe a better name is in order?

Reply:
Whoops, just realized you said `vec.clear()` not `bump.clear()`. This is a good point

Reply:
clear drops the contents, but the contents don't have a drop so that's a no-op. Interestingly `Vec::clear()` promises:
> Note that this method has no effect on the allocated capacity of the vector.

We should just use clear().